doc/build/flashing/configuration.rst

```diff
@@ -0,0 +1,92 @@
+:ref:`west flash<west-flashing>`) which allows for customising how commands are used when
+programming boards. This configuration is used for :ref:`sysbuild` projects and allows for
+configuring when commands are ran for groups of board targets. As an example: a multi-core SoC
+might want to only allow the ``--erase`` argument to be used once for all of the cores in the SoC
+which would prevent multiple erase tasks running in a single ``west flash`` invocation, which
+could wrongly clear the memory which is used by the other images being programmed.
+
+Priority
+********
+
+Flashing configuration is singular, it will only be read from a single location, this
+configuration can reside in the following files starting with the highest priority:
+
+ * ``soc.yml`` (in soc folder)
+ * ``board.yml`` (in board folder)
+
+Configuration
+*************
+
+Configuration is applied by using a ``runners`` map with a ``run_once`` child, from here is a list
+of command as they would be provided to the flash runner e.g. ``--reset`` and the settings for
+each of these commands. Commands have associated runners that they apply to using a ``runners``
+list value, this can contain ``all`` if it applies to all runners, otherwise must contain each
+runner that it applies to using the runner-specific name. Groups of board targets can be specified
+using the ``groups`` key which has a list of board target sets - board target are regular
+expression matches, for ``soc.yml`` files each set of board targets must be in a ``qualifiers`` key
+(only regular expression matches for board qualifiers are allowed) and for ``board.yml`` files each
+set of board targets must be in a ``boards`` key, these are lists containing the matches which form
+a singular group. A final parameter ``run`` can be set to ``first`` which means that the command
+will be ran once with the first image flashing process per set of board targets, or to ``last``
+which will be ran once for the final image flash per set of board targets.
+
+An example flashing configuration for a ``soc.yml`` is shown below in which the ``--recover``
+command will only be used once for any board targets which used the nRF5340 SoC application or
+network CPU cores, and will only reset the network or application core after all images for the
+respective core have been flashed.
+
+.. code-block:: yaml
+
+  runners:
+    run_once:
+      - '--recover':
+          run: first
+          runners:
+            - nrfjprog
+          groups:
+            - qualifiers:
+                - (.+)/nrf5340/cpunet
+                - (.+)/nrf5340/cpuapp
+                - (.+)/nrf5340/cpuapp/ns
+      - '--reset':
+          run: last
+          runners:
+            - nrfjprog
+            - jlink
+          groups:
+            - qualifiers:
+                - (.+)/nrf5340/cpunet
+            - qualifiers:
+                - (.+)/nrf5340/cpuapp
+                - (.+)/nrf5340/cpuapp/ns
```

Comment:
I discussed this internally with @nordicjm and @tejlmand but I will post my opinion on it here.
My problem with this approach is that you are essentially using the following format, translating the `runners` node into pseudo-Python:

`{run_once: [{'--recover': {run: ... , runners: ...}}, {'--reset': {}} ]}`

so it's a map, and the first key is `run_once`. Then, the value corresponding to the key `run_once` is what I was an array of dictionaries, where each of these dictionaries has a single key which is the name of the command (`--recover`, `--reset`). The value of that key then is in turn a map with a bunch of keys (`run`, `runners`, `groups`, ...).
I personally dislike this approach because you end up with an array of single-key/value dictionaries. Your rationale for this choice is that snippets currently use this format, so it's a matter of consistency. But I partially disagree with that statement, because the snippet .yaml files use a significantly different approach: instead of having an array of single-key/value dictionaries, snippets use a single dictionary in which each key is a regular expression. I think that the snippet approach is valid, because it uses the dictionary data structure in a way that (to me at least) makes sense: filtering keys and then looking up the value. You pointed out that the exact same approach that snippets use is not possible here, due to the fact that, in the `run_once` case, the commands can appear more than once, so they cannot be dictionary keys, hence the use of an array.
Now, in that case, and since we have an array anyway, I do prefer @tejlmand's proposal of an array of dictionaries, where each dictionary has multiple key/value pairs and one of which is then `command:` (key) `--recover` (value). Because I think it makes more sense to use an array in this fashion, and also reduces the nesting by one level. With this approach you get, again in Python pseudo-code:

`{run_once: [{command: '--recover', run: ... , runners: ...}, {command: '--reset', run: ..., runners: ...} ]}`

To make my point about the nesting levels a bit more clear, I'd like to compare accessing an item in both proposals. I will compare the `run` dict item to `first`:

In your initial proposal: 
`run_once[i]['--recover'][run] == first`
where `i` is the item index in the `run_once` sequence that has `--recover` as its dict key

Alternative approach:
`run_once[i][run] == first`
where `i` is the item index in the `run_once` sequence that has `--recover` as the value of its command key

Reply:
Well the whole format sounds a bit weird, the whole "run once" -> "first/last" thing. Would it make sense to instead have something like a "pre_flash {command, runner...}" and "post_flash {command, runner...}"? Anything to make this a bit more compact as well.

Reply:
I don't think so @fabiobaltieri. The nature of the combinations needed requires a system like this, because you need to be able to specify even multiple times a single command to be run once. pre_flash and post_flash just wouldn't be flexible enough :(

Reply:
I've now seen the latest proposal, which effectively translates to:

`{run_once: {'--recover': [{run: ... , runners: ...}], '--reset': [{run: ..., runners: ...}, {run: ..., runners: ...}] }}`

This indeed uses a single dict with the different commands as the keys, which is much closer to what snippets do, and much cleaner from my PoV.